File: bank-account-project/src/main/java/com/example/bank_account_project/service/UserService.java
package com.example.bank_account_project.service;

import com.example.bank_account_project.dto.UserDto;
import com.example.bank_account_project.entity.User;
import com.example.bank_account_project.utils.validable.Validable;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Transactional
public interface UserService {

    Validable<UserDto> createUser(final UserDto userDto);

    Optional<UserDto> getUserById(final Long id);

    List<UserDto> searchUsers(final String search);

    Validable<UserDto> updateUser(final Long id, final UserDto userDto);

    void deleteUser(final Long id);

    User updateAccountBalance(final Long userId, final Double transactionAmount);
}

File: web-app-angular/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { UserDto } from '../models/user.dto';

@Injectable({
  providedIn: 'root',
})
export class UserService {
  private apiUrl = 'http://localhost:8080';

  constructor(private http: HttpClient) {}

  searchUsers(search: string): Observable<any> {
    const params = new FormData();
    params.append('search', search);

    return this.http.post<any>(`${this.apiUrl}/user/search-user`, params);
  }

  getUserById(userId: string): Observable<UserDto> {
    return this.http.get<UserDto>(`${this.apiUrl}/${userId}`);
  }

  createUser(user: UserDto): Observable<any> {
    return this.http.post<any>(this.apiUrl, user);
  }

  updateUser(userId: string, user: UserDto): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/${userId}`, user);
  }

  deleteUser(userId: string): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${userId}`);
  }
}
